api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,254 @@
+
+  // Change how and if the filter waits for responses from the server
+  // for future messages. Servers may use this to switch between
+  // synchronous and asynchronous processing modes based on what they see.
+  // A message with this field may be sent even if none of the "oneof" fields
+  // have been sent, and it may be sent out-of-band with respect to
+  // the other messages if desired.
+  ProcessingModeChange mode_change = 10;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. If multiple headers share the same key, they
+  // must be merged according to the HTTP spec. All header keys must be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool no_body = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a RequestHeaders or ResponseHeaders
+// message.
+message HeaderResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+}
+
+message BodyResponse {
+  // Add or replace HTTP headers.
+  repeated config.core.v3.HeaderValueOption set_headers = 1;
+
+  // Remove these HTTP headers.
+  repeated string remove_headers = 2;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 3;
+}
+
+// This message completely replaces the response.
+// It will immediately return the data specified here to the downstream
+// system and stop further filter processing.
+message ImmediateResponse {
+  type.v3.HttpStatus status = 1 [(validate.rules).message = {required: true}];
+
+  repeated config.core.v3.HeaderValue headers = 2;
+
+  bytes body = 3;
+
+  repeated config.core.v3.HeaderValue trailers = 4;
+}
+
+// Change all or part of the message body by replacing the portion
+// of the last chunk received denoted by "offset" and "length" with "body."
+message BodyMutation {
```

Comment:
can we think offhand of any case where we want to modify a few bytes?  I'm used to most filters passing through the body, and a very few doing transformation (usually compression or decompression).  If there are delta-use cases please ignore this comment but I wonder if we're just better off saying the server can either have no effect on body or send its own response body.

Reply:
I think that replacing the whole body, in most cases as one single buffer, will be pretty common in the kinds of use cases I'm used to from the API management and integration world. For instance, transform a message body from one format to another (JSON to XML to protobuf and so on), or run something like XSLT or JavaScript on the message body for a larger transformation.

So yes, I think that replacing the whole body will be pretty common.

Modifying a few bytes is going to be pretty uncommon, I agree, and I could see the need to just replace the whole body, which would eliminate a set of errors.

This bit is supposed to work like the WASM ABI, which lets a filter append or prepend or modify the middle. I have mixed feelings about whether or not we need that complexity here.